File: src/main/java/edu/neu/coe/info6205/sort/simple/SelectionSort.java
/*
  (c) Copyright 2018, 2019 Phasmid Software
 */
package edu.neu.coe.info6205.sort.simple;

import edu.neu.coe.info6205.sort.BaseHelper;
import edu.neu.coe.info6205.sort.Helper;
import edu.neu.coe.info6205.sort.SortWithHelper;
import edu.neu.coe.info6205.util.Config;

public class SelectionSort<X extends Comparable<X>> extends SortWithHelper<X> {

    /**
     * Constructor for SelectionSort
     *
     * @param N      the number elements we expect to sort.
     * @param config the configuration.
     */
    public SelectionSort(int N, Config config) {
        super(DESCRIPTION, N, config);
    }

    public SelectionSort() {
        this(new BaseHelper<>(DESCRIPTION));
    }

    /**
     * Constructor for SelectionSort
     *
     * @param helper an explicit instance of Helper to be used.
     */
    public SelectionSort(Helper<X> helper) {
        super(helper);
    }

    public void sort(X[] xs, int from, int to) {
        final Helper<X> helper = getHelper();
        // TO BE IMPLEMENTED
     }

    /**
     * This is used by unit tests.
     *
     * @param ys  the array to be sorted.
     * @param <Y> the underlying element type.
     */
    public static <Y extends Comparable<Y>> void mutatingSelectionSort(Y[] ys) {
        new SelectionSort<Y>().mutatingSort(ys);
    }

    public static final String DESCRIPTION = "Selection sort";

}

File: Python/src/sort/simple/selection_sort.py
from typing import List

from sort.simple.helper import Helper
from sort.simple.sort import Sort
from util.generic_type import C


class SelectionSort(Sort[C]):

    def __init__(self, helper: Helper[C] = None):
        if helper is None:
            helper = Helper[C]("SelectionSort")
        self.helper = helper

    def _sort(self, xs: List[C], _from: int, _to: int) -> None:
        
        # TO BE IMPLEMENTED ...
        
        for i in range(_from, _to):
            # Invariant 1: xs[from..i] are in (final) ascending order
            # Invariant 2: no element xs[i+1..to-1] is smaller than any element xs[from..i]

        # ... END IMPLEMENTATION
           

    def get_helper(self) -> Helper[C]:
        return self.helper
